### ISSUE TITLE
Support Govcloud & China Partitions

### ISSUE THREAD

**Issue by richardcase** [kubernetes-sigs/cluster-api-provider-aws#4216]
/kind feature
/help
/triage accepted
/priority important-longterm

**Describe the solution you'd like**
We should be able to support using CAPA in other AWS partitions such as Govcloud and China. Currently, we still assume the `aws` partition in our code in certain places. For example:

- https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/c42dc926cdb75cba82558d71a0341951a19cc085/pkg/cloud/services/eks/roles.go#L42
- https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/c42dc926cdb75cba82558d71a0341951a19cc085/pkg/cloud/services/s3/s3.go#L245

**clusterawsadm** allows you to change the partition via a confile file:

```yaml
apiVersion: bootstrap.aws.infrastructure.cluster.x-k8s.io/v1beta1
kind: AWSIAMConfiguration
spec:
  partition: aws-us-gov
```

**Anything else you would like to add:**

This is about going through the code and ensuring that we don't assume the **aws** partition. From the controller manager, we should be able to get the partition from the creds of the user.

**Environment:**

- Cluster-api-provider-aws version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

**Comment by wilderk** [comment]
To add an example, we are deploying in AWS GovCloud. The iam cloudformation fails because the GovCloud partition is "aws-us-gov".
```
clusterawsadm bootstrap iam create-cloudformation-stack
  Attempting to create AWS CloudFormation stack cluster-api-provider-aws-sigs-k8s-io
  Error: failed to create AWS CloudFormation stack: ResourceNotReady: failed waiting for successful resource state
```
Error in AWS console
```
Partition "aws" is not valid for resource "arn:aws:iam::*:role/aws-service-role/eks-fargate-pods.amazonaws.com/AWSServiceRoleForAmazonEKSForFargate". (Service: AmazonIdentityManagement; Status Code: 400; Error Code: MalformedPolicyDocument; Request ID: aae7ef41-a526-4dad-8156-e99eb6be9e8b; Proxy: null)
```
Setting config did not solve the issue.
```
# create an iam boostrap config file
cat << EOF > bootstrap-config.yaml
apiVersion: bootstrap.aws.infrastructure.cluster.x-k8s.io/v1beta1
kind: AWSIAMConfiguration
spec:
  partition: aws-us-gov
EOF

# create the iam CF stack
clusterawsadm bootstrap iam create-cloudformation-stack --config bootstrap-config.yaml
  Attempting to create AWS CloudFormation stack cluster-api-provider-aws-sigs-k8s-io
  Error: failed to create AWS CloudFormation stack: ResourceNotReady: failed waiting for successful resource state
```

**Comment by wilderk** [comment]
Yes. Please try that PR again.

**Comment by wilderk** [comment]
This issue is resolved with https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4010

**Comment by richardcase** [comment]
As per the comment, this is  closed in #4010 

/close
